libraries/AP_HAL_Linux/CAN.h

```diff
@@ -0,0 +1,213 @@
+
+    bool hasReadyTx() const;
+
+    bool hasReadyRx() const;
+
+    void poll(bool read, bool write);
+
+    int16_t configureFilters(const uavcan::CanFilterConfig* filter_configs, uint16_t num_configs) override;
+
+    uint16_t getNumFilters() const override;
+
+    uint64_t getErrorCount() const override;
+
+
+private:
+    struct TxItem
+    {
+        uavcan::CanFrame frame;
+        uavcan::MonotonicTime deadline;
+        uavcan::CanIOFlags flags = 0;
+        std::uint64_t order = 0;
+
+        TxItem(const uavcan::CanFrame& arg_frame, uavcan::MonotonicTime arg_deadline,
+               uavcan::CanIOFlags arg_flags, std::uint64_t arg_order)
+            : frame(arg_frame)
+            , deadline(arg_deadline)
+            , flags(arg_flags)
+            , order(arg_order)
+        { }
+
+        bool operator<(const TxItem& rhs) const
+        {
+            if (frame.priorityLowerThan(rhs.frame)) {
+                return true;
+            }
+            if (frame.priorityHigherThan(rhs.frame)) {
+                return false;
+            }
+            return order > rhs.order;
+        }
+    };
+
+    struct RxItem
+    {
+        uavcan::CanFrame frame;
+        uavcan::MonotonicTime ts_mono;
+        uavcan::UtcTime ts_utc;
+        uavcan::CanIOFlags flags;
+
+        RxItem()
+            : flags(0)
+        { }
+    };
+
+    void _pollWrite();
+
+    void _pollRead();
+
+    int _write(const uavcan::CanFrame& frame) const;
+
+    int _read(uavcan::CanFrame& frame, uavcan::UtcTime& ts_utc, bool& loopback) const;
+
+    void _incrementNumFramesInSocketTxQueue();
+
+    void _confirmSentFrame();
+
+    bool _wasInPendingLoopbackSet(const uavcan::CanFrame& frame);
+
+    bool _checkHWFilters(const can_frame& frame) const;
+
+    void _registerError(SocketCanError e) { _errors[e]++; }
+
+    uint32_t _bitrate;
+
+    bool _initialized;
+
+    int _fd;
+
+    const unsigned _max_frames_in_socket_tx_queue;
+    unsigned _frames_in_socket_tx_queue;
+    uint64_t _tx_frame_counter;
+
+    std::map<SocketCanError, uint64_t> _errors;
+    std::priority_queue<TxItem> _tx_queue;
+    std::queue<RxItem> _rx_queue;
+    std::unordered_multiset<uint32_t> _pending_loopback_ids;
+    std::vector<can_filter> _hw_filters_container;
+};
+
+class LinuxCANDriver: public AP_HAL::CANManager {
+public:
+    static LinuxCANDriver *from(AP_HAL::CANManager *can)
+    {
+        return static_cast<LinuxCANDriver*>(can);
+    }
+
+    LinuxCANDriver() : _initialized(false) { _ifaces.reserve(uavcan::MaxCanIfaces); }
+    ~LinuxCANDriver() { }
+
+    void register_thread();
+
+    void _timer_tick();
+
+    bool begin(uint32_t bitrate, uint8_t can_number) override;
+
+    bool is_initialized() override;
+
+    void initialized(bool val) override;
+
+    AP_UAVCAN *p_uavcan;
+
+    AP_UAVCAN *get_UAVCAN(void) override;
+
+    void set_UAVCAN(AP_UAVCAN *uavcan) override;
+
+    virtual LinuxCAN* getIface(uint8_t iface_index) override;
```

Comment:
is this override correct? I can't see a method in AP_HAL returning something of LinuxCAN type.

Reply:
Well, this is slightly incorrect. This has to return an ICanIface* type. The method is not defined in AP_HAL, this is defined in uavcan module [here](https://github.com/ArduPilot/uavcan/blob/9d75eba870e3b3935346229267ecee2c499f6575/libuavcan/include/uavcan/driver/can.hpp#L211).